kernel/src/last_checkpoint_hint.rs

```diff
@@ -0,0 +1,65 @@
+//! Utities for reading the `_last_checkpoint` file. Maybe this file should instead go under
+//! log_segment module since it should only really be used there? as hint for listing?
+
+use crate::schema::Schema;
+use crate::{DeltaResult, Error, StorageHandler, Version};
+use delta_kernel_derive::internal_api;
+
+use serde::{Deserialize, Serialize};
+use tracing::warn;
+use url::Url;
+
+/// Name of the _last_checkpoint file that provides metadata about the last checkpoint
+/// created for the table. This file is used as a hint for the engine to quickly locate
+/// the latest checkpoint without a full directory listing.
+pub(crate) const LAST_CHECKPOINT_FILE_NAME: &str = "_last_checkpoint";
+
+// Note: Schema can not be derived because the checkpoint schema is only known at runtime.
+#[derive(Debug, Deserialize, Serialize, PartialEq, Eq)]
+#[serde(rename_all = "camelCase")]
+#[internal_api]
+pub(crate) struct LastCheckpointHint {
+    /// The version of the table when the last checkpoint was made.
+    #[allow(unreachable_pub)] // used by acceptance tests (TODO make an fn accessor?)
+    pub version: Version,
+    /// The number of actions that are stored in the checkpoint.
+    pub(crate) size: i64,
+    /// The number of fragments if the last checkpoint was written in multiple parts.
+    pub(crate) parts: Option<usize>,
+    /// The number of bytes of the checkpoint.
+    pub(crate) size_in_bytes: Option<i64>,
+    /// The number of AddFile actions in the checkpoint.
+    pub(crate) num_of_add_files: Option<i64>,
+    /// The schema of the checkpoint file.
+    pub(crate) checkpoint_schema: Option<Schema>,
+    /// The checksum of the last checkpoint JSON.
+    pub(crate) checksum: Option<String>,
+}
+
+impl LastCheckpointHint {
+    /// Try reading the `_last_checkpoint` file.
+    ///
+    /// Note that we typically want to ignore a missing/invalid `_last_checkpoint` file without
+    /// failing the read. Thus, the semantics of this function are to return `None` if the file is
+    /// not found or is invalid JSON. Unexpected/unrecoverable errors are returned as `Err` case and
+    /// are assumed to cause failure.
+    // TODO(#1047): weird that we propagate FileNotFound as part of the iterator instead of top-
+    // level result coming from storage.read_files
+    pub(crate) fn read(
```

Comment:
Would a `try_` prefix helpful, given its signature?
```suggestion
    pub(crate) fn try_read(
```